xugrid/regrid/structured.py

```diff
@@ -21,56 +28,619 @@ class StructuredGrid1d:
 
-    def length(self):
-        return abs(np.diff(self.bounds, axis=1))
+    def locate_centroids(self, other: "StructuredGrid1d"):
+        """
+        Returns source and target indexes based on nearest neighbor of centroids
 
+        Args:
+            self (StructuredGrid1d): source grid
+            other (StructuredGrid1d): target grid
 
-class StructuredGrid2d:
+        Returns:
+            source_index (np.array): source indexes
+            target_index (np.array): target indexes
+            weights (np.array): array of ones
+        """
+        source_index, target_index = self.valid_nodes_index(other)
+        weights = np.ones(source_index.size, dtype=float)
+        return self.sorted_output(source_index, target_index, weights)
+
+    def linear_weights(self, other: "StructuredGrid1d"):
+        """
+        Returns linear source and target indexes and corresponding weights
+
+        Args:
+            self (StructuredGrid1d): source grid
+            other (StructuredGrid1d): target grid
+
+        Raises:
+            ValueError: when number of nodes is to small to compute linear weights
+
+        Returns:
+            source_index (np.array): overlaying self indexes
+            target_index (np.array): overlaying other indexes
+            weights (np.array): array linear weights
+        """
+
+        source_index, target_index = self.valid_lineair_nodes_index(other)
+        weights = self.compute_weights(other, source_index, target_index)
+        source_index, target_index, weights = self.centroids_to_linear_indexes(
+            other, source_index, target_index, weights
+        )
+        return self.sorted_output(source_index, target_index, weights)
+
+
+class StructuredGrid2d(StructuredGrid1d):
     """
     e.g. (x,y) -> (x,y)
-
-    Parameters
-    ----------
-    xbounds: (nx, 2)
-    ybounds: (ny, 2)
     """
 
     def __init__(
         self,
-        xbounds,
-        ybounds,
+        obj: Union[xr.DataArray, xr.Dataset],
+        name_x: str,
+        name_y: str,
     ):
-        self.xbounds = StructuredGrid1d(xbounds)
-        self.ybounds = StructuredGrid1d(ybounds)
+        self.xbounds = StructuredGrid1d(obj, name_x)
+        self.ybounds = StructuredGrid1d(obj, name_y)
+
+    @property
+    def ndim(self):
+        return 2
+
+    @property
+    def dims(self):
+        return self.ybounds.dims + self.xbounds.dims  # ("y", "x")
+
+    @property
+    def size(self):
+        return self.ybounds.size * self.xbounds.size
 
     @property
     def shape(self):
         return (self.ybounds.size, self.xbounds.size)
 
+    @property
+    def area(self):
+        return np.multiply.outer(self.ybounds.length, self.xbounds.length)
+
+    def convert_to(self, matched_type):
+        if isinstance(self, matched_type):
+            return self
+        elif isinstance(UnstructuredGrid2d):
+            return Ugrid2d.from_structured(self.xbounds, self.ybounds)
```

Comment:
Its there because users could enter a structured source and unstructured target for example. Then it works like this:
Initiate regridder (source = xarray, target = ugrid)
It goes to the BaseRegridder which sets up the grids via setup_grid(source), setup_grid(target). So in this case xarray = StructuredGrid2d and ugrid = UnstructuredGrid2d
Then in the specific regridder (linear for example) we will match(source, target) based on promotions:
For this case we use the rule: structured, unstructured = unstructured
so the source will (via match) enter the StructuredGrid2d class via source.convert_to(matched_type =UnstructuredGrid2d). Then in the structured class is converted to unstructured and returned. And from there on goes via the unstructured path. In the unstructured class, the same logic is implemented.

Reply:
I will fix line 357 indeed